lib/spack/spack/package.py

```diff
@@ -1628,7 +1643,9 @@ def build_process():
                     echo = logger.echo
                     self.log()
 
-                # Run post install hooks before build stage is removed.
+                # Run post install hooks for the code before build stage is
+                # removed.  This step requires stage.source_path and install
```

Comment:
Questions:

* Does `stage.source_path` not exist at some point when the stage exists? In particular at this point when the install has been conducted. In other words unless I'm missing something I think `Stage.source_path` would always exist at this point
* What enforces that the 'install' phase exists if we get to this point?

I'm thinking these notes are to distinguish no-source packages. Although I think (a) as noted in the `spack.build_environment.fork` call in my review I thought we'd want to make it possible to have an install phase and (b) even without any phases I think we would at least want to be able to generate a module for a bundle package (which is a post-install hook)

Reply:
Restored original comment.

Reply:
> Questions:
> 
>     * Does `stage.source_path` not exist at some point when the stage exists? In particular at this point when the install has been conducted. In other words unless I'm missing something I think `Stage.source_path` would always exist at this point

Not for no-source packages since `stage.source_path` is created when the fetch strategy expands the archive. 

> 
>     * What enforces that the 'install' phase exists if we get to this point?
> 

Nothing for `BundlePackage` at this point.

> 
> I'm thinking these notes are to distinguish no-source packages. Although I think (a) as noted in the `spack.build_environment.fork` call in my review I thought we'd want to make it possible to have an install phase and (b) even without any phases I think we would at least want to be able to generate a module for a bundle package (which is a post-install hook)

Okay.  It looks like the following are required in order to enable the current `nosource` test to get through the `build_process`:
- `pkg.do_stage` should create the `stage.source_path` for no-source packages
- `BundlePackage` must have a `phases` property (i.e., empty list)

However, those changes are not sufficient for a new test case that defines a post-install hook.

Reply:
Finished resolving issues no-source install+post-install support.  The new test case works with `mock_fetch`, which by-passes the new fetch strategy but was necessary for the `nosource` package's dependencies to "install" correctly.

Reply:
The latest change includes removal of the `NoFetchStrategyError` given the introduction of a `BundleFetchStrategy`.

Additionally, testing has confirmed the following when not using the `mock_fetch` fixture:
- Using the `BundleFetchStrategy` results in `FetchError: All fetchers failed...` for the `nosource*` packages; and
- Not using the `BundleFetchStrategy` results in `InvalidArgsError: Could not guess a fetch strategy`.